### ISSUE TITLE
Minimal Version of Json.NET

### ISSUE THREAD

**Issue by shytikov** [JamesNK/Newtonsoft.Json#707]
Hi,

Have anyone considered releasing stripped-down version of `Json.NET` that has only basic functionality?

I'm developing plugins for closed-source third-party system that runs in partial trust in addition to all kind of restrictions. For many years I have used custom written ugly `JSON` serializers, since `Json.NET` did not worked in partial trust environments successfully.

Not any more. Version 7.0.1 finally passes my tests and we can use it in production. In sandboxed mode. But. We ship our plugins merged with all required libraries. And adding one `Json.NET` assembly will increase weight of the package 10 times. Which is very hard to explain to our customers. 

So my question is, can you help me identify parts of `Json.NET` that could be de-coupled from main, core functionality (serialization-deserialization) and shipped as dependency? Which parts of JSON.NET takes most of the library size?

I have asked same question on SO: http://stackoverflow.com/q/33564729/1047741 but there people seems to be live in reality where everyone who has different needs looks stupid. 

So far I managed to strip down BSON and save 50 Kb of library size. But this is not enough...

Thank you so much in advance for your (P.C.) help!

**Comment by shytikov** [comment]
Here is results of my experiments. 

* Stripping down `BSON` gives around 50 Kb of extra space;
* Stripping down `JPath` gives around 20 Kb of extra space;
* Stripping down `Schema` gives another 50 Kb.

But this is not enough it's still 450 Kb library... 

`Json.NET` Experts can you help me with the list of functionality, that could also be decoupled (if I don't need it)?

**Comment by mburbea** [comment]
JSON.Net 7.0.1 dll is <500kb what environment are you in that 500kb will cause that much concern?

**Comment by JamesNK** [comment]
I might eventually move the JsonReader/JsonWriter into its own package. And maybe the BSON stuff into its own package. No plans to do anything in the short term. I would wait until project.json is more widespread.

**Comment by shytikov** [comment]
@mburbea it's MS Dynamics CRM Online. Customization of this system is done using plugins running in partial trust.

There is no access to GAC or filesystem. The only way how plugins are installed is upload database. If plugin dependent on another third-party library, there is no other way to make it work, expect package dependency together with main plugin using `ILMerge`, for example. 

Plugins are totally isolated from each other, so if another plugin needs `Json.NET`, which was already packaged with first one, there is no other way to make second plugin work, except package `Json.NET` with it too.

Now imagine, that complex and loaded system has large number of plugins (which normally is the case) each packaged with half megabyte `Json.NET`... Ok, I don't count database space waisted, but think about memory, and performance loading / unloading and executing such plugins in environment which is used by thousands of users and each execution is triggered in separate sandbox.

It's a bit frustrating. Before we used some custom written JSON serializers / deserializers, but they are not as good as `Json.NET`, and in addition they also might have some issues working in partial trust (they fails in specific cases, that requires to handle them separately).

But newest release of `Json.NET` actually changed score. With surprise I found out that it not fails in serialization / deserialization tasks in partial trust anymore.

So it would be nice to use it. But this half of the megabyte size really noticeable, when it comes to loaded environment.

**Comment by shytikov** [comment]
@JamesNK yes, extracting of `JsonReader` and `JsonWriter` it's basically what I'm looking for...

Will try to accomplish it by myself. Do you have any idea (as an author) will it be worth it? I mean, if resulting `JsonReader` / `JsonWriter` will be 399 Kb it's not that big difference to fight for... Through if it would be 100 Kb or less, it definitely worth it.

**Comment by shytikov** [comment]
You can see what I have achieved so far here: https://github.com/JamesNK/Newtonsoft.Json/compare/master...shytikov:master

**Comment by veblush** [comment]
@shytikov FYI, I ported Json.Net 8 to Unity3D ([Json.Net.Unity3D](https://github.com/SaladbowlCreative/Json.Net.Unity3D))

Two version is ported. Regular one and Lite one. Lite one stripped many features such as bson. json.linq, jpath, schema, ..., etc. The size of DLL is 243KB which is almost half of original DLL.

[Assembly Size](https://github.com/SaladbowlCreative/Json.Net.Unity3D/blob/master/NOTE.md) shows more details of each versions in sections of DLL. Interesting thing is that whole IL code accounts for just 1/3 of DLL and meta data is quite larger than expected. (a number of classes, methods, field, property can bloat meta data)

**Comment by shytikov** [comment]
@veblush thanks! It looks like what I was searching for!
